source/gameengine/Rasterizer/RAS_CubeMap.cpp

```diff
@@ -0,0 +1,161 @@
+#include "GPU_framebuffer.h"
+#include "GPU_draw.h"
+
+#include "DNA_texture_types.h"
+
+#include "glew-mx.h"
+
+static const MT_Matrix4x4 bottomFaceViewMat(
+	-1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 1.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 topFaceViewMat(
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, -1.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 rightFaceViewMat(
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 1.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 leftFaceViewMat(
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, -1.0f, 0.0f,
+	0.0f, 1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 backFaceViewMat(
+	0.0f, 0.0f, 1.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 frontFaceViewMat(
+	0.0f, 0.0f, -1.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	-1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+
+MT_Matrix4x4 RAS_CubeMap::facesViewMat[6] = {topFaceViewMat, bottomFaceViewMat, frontFaceViewMat, backFaceViewMat, rightFaceViewMat, leftFaceViewMat};
+MT_Matrix3x3 RAS_CubeMap::camOri[6] = {
+	topFaceViewMat.to3x3(),
+	bottomFaceViewMat.to3x3(),
+	frontFaceViewMat.to3x3(),
+	backFaceViewMat.to3x3(),
+	leftFaceViewMat.to3x3(),
+	rightFaceViewMat.to3x3()
+};
+
+static const GLenum cubeMapTargets[6] = {
+	GL_TEXTURE_CUBE_MAP_POSITIVE_Z_ARB,
+	GL_TEXTURE_CUBE_MAP_NEGATIVE_Z_ARB,
+	GL_TEXTURE_CUBE_MAP_POSITIVE_X_ARB,
+	GL_TEXTURE_CUBE_MAP_NEGATIVE_X_ARB,
+	GL_TEXTURE_CUBE_MAP_POSITIVE_Y_ARB,
+	GL_TEXTURE_CUBE_MAP_NEGATIVE_Y_ARB,
+};
+
+RAS_CubeMap::RAS_CubeMap(RAS_Texture *texture, RAS_IRasterizer *rasty)
+	:m_cubeMapTexture(NULL),
+	m_useMipmap(false),
+	m_texture(texture)
+{
+	EnvMap *env = m_texture->GetMTex()->tex->env;
+
+	m_useMipmap = (env->flag & ENVMAP_MIPMAP) != 0;
+
+	m_cubeMapTexture = m_texture->GetGPUTexture();
+	// Increment reference to make sure the gpu texture will not be freed by someone else.
+	GPU_texture_ref(m_cubeMapTexture);
+
+	if (!m_useMipmap) {
+		// Disable mipmaping.
+		GPU_texture_bind(m_cubeMapTexture, 0);
+		GPU_texture_filter_mode(m_cubeMapTexture, false, false);
+		GPU_texture_unbind(m_cubeMapTexture);
+	}
+
+	for (unsigned short i = 0; i < 6; ++i) {
+		m_fbos[i] = GPU_framebuffer_create();
+		m_rbs[i] = GPU_renderbuffer_create(GPU_texture_width(m_cubeMapTexture), GPU_texture_height(m_cubeMapTexture),
+										   0, GPU_HDR_NONE, GPU_RENDERBUFFER_DEPTH, NULL);
+
+		GPU_framebuffer_texture_attach_target(m_fbos[i], m_cubeMapTexture, cubeMapTargets[i], 0, NULL);
+		GPU_framebuffer_renderbuffer_attach(m_fbos[i], m_rbs[i], 0, NULL);
+	}
+}
+
+RAS_CubeMap::~RAS_CubeMap()
+{
+	GPU_texture_free(m_cubeMapTexture);
+	GPU_free_image(m_texture->GetImage());
+
+	for (unsigned short i = 0; i < 6; ++i) {
+		GPU_framebuffer_free(m_fbos[i]);
+		GPU_renderbuffer_free(m_rbs[i]);
+	}
+}
+
+void RAS_CubeMap::BeginRender()
```

Comment:
Why ? it freed the frame buffers and render buffers created by the cube map, also freeing the texture/image just ask to the next user of this texture to reload it.